test/Microsoft.AspNetCore.Mvc.Formatters.Json.Test/Internal/JsonResultExecutorTest.cs

```diff
@@ -217,6 +219,57 @@ public async Task ExecuteAsync_NullResult_LogsNull()
             Assert.Equal(expected, logger.MostRecentMessage);
         }
 
+        [Fact]
+        public async Task ExecuteAsync_WritesToTheResponseStream_WhenContentIsLargerThanBuffer()
+        {
+            // Arrange
+            var writeLength = 2 * TestHttpResponseStreamWriterFactory.DefaultBufferSize + 4;
+            var text = new string('a', writeLength);
+            var expectedWriteCallCount = Math.Ceiling((double)writeLength / TestHttpResponseStreamWriterFactory.DefaultBufferSize);
+
+            var stream = new Mock<Stream>();
+            stream.SetupGet(s => s.CanWrite).Returns(true);
+            var httpContext = new DefaultHttpContext();
+            httpContext.Response.Body = stream.Object;
+            var actionContext = new ActionContext(httpContext, new RouteData(), new ActionDescriptor());
+
+            var result = new JsonResult(text);
+            var executor = CreateExecutor();
+
+            // Act
+            await executor.ExecuteAsync(actionContext, result);
+
+            // Assert
+            stream.Verify(s => s.Write(It.IsAny<byte[]>(), It.IsAny<int>(), TestHttpResponseStreamWriterFactory.DefaultBufferSize), Times.Exactly(2));
+            stream.Verify(s => s.Flush(), Times.Never());
+            stream.Verify(s => s.WriteAsync(It.IsAny<byte[]>(), It.IsAny<int>(), It.IsAny<int>(), It.IsAny<CancellationToken>()), Times.Once());
+        }
+
+        [Theory]
+        [InlineData(5)]
+        [InlineData(TestHttpResponseStreamWriterFactory.DefaultBufferSize - 30)]
+        public async Task ExecuteAsync_DoesNotWriteSynchronouslyToTheResponseBody_WhenContentIsSmallerThanBufferSize(int writeLength)
+        {
+            // Arrange
+            var text = new string('a', writeLength);
+
+            var stream = new Mock<Stream>();
+            stream.SetupGet(s => s.CanWrite).Returns(true);
+            var httpContext = new DefaultHttpContext();
+            httpContext.Response.Body = stream.Object;
+            var actionContext = new ActionContext(httpContext, new RouteData(), new ActionDescriptor());
+
+            var result = new JsonResult(text);
+            var executor = CreateExecutor();
+
+            // Act
+            await executor.ExecuteAsync(actionContext, result);
+
+            // Assert
+            stream.Verify(s => s.Flush(), Times.Never());
+            stream.Verify(s => s.Write(It.IsAny<byte[]>(), It.IsAny<int>(), It.IsAny<int>()), Times.Never());
```

Comment:
I think it's worth a comment explaining how these verifications are meaningful. I know this area and I had to think about it for 30s.